### PR TITLE
CONSOLE-3383: Minimal Dynamic Plugin for the CronTab CRD

### PR DESCRIPTION
**Pull request by raspbeep**
Adds minimal dynamic plugin for CronTab CRD.

https://issues.redhat.com/browse/CONSOLE-3383

Signed-off-by: Pavel Kratochvil <pakratoc@redhat.com>

### CODE REVIEW

**Review by jcaianirh — CHANGES_REQUESTED** [review]
@raspbeep I looked through the readme to install and there are refs to helm installs. I think the install section was taken from another plugin and needs to be cleaned up.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+`--platform=linux/amd64` when building the image to target the correct platform
+to run in-cluster.
+
+## Deployment on cluster
+
+After pushing an image with your changes to an image registry, you can deploy
+the plugin to a cluster by instantiating the template:
+
+```sh
+oc process -f oc-manifest.yaml \
```
Comment (spadgett) [inline README.md:82]: This looks wrong. oc-manifest.yaml is not a template, so you shouldn't use `oc process`.

Is there a reason not to use the helm chart from the template repository?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+2. `(Ctrl+Shift+P) => Remote Containers: Open Folder in Container...`
+3. `yarn dev`
+4. Navigate to <http://localhost:9000>
+
+#### Cypress Testing inside the container
+
+1. `yarn test-cypress-docker`
+2. Navigate to <http://localhost:10000>
+3. login with password `kubevirt` (no need for username)
```
Comment (spadgett) [inline README.md:41]: Why is kubevirt the password?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+OC_PASS=<password>
+```
+
+2. `(Ctrl+Shift+P) => Remote Containers: Open Folder in Container...`
+3. `yarn dev`
+4. Navigate to <http://localhost:9000>
+
+#### Cypress Testing inside the container
+
+1. `yarn test-cypress-docker`
```
Comment (spadgett) [inline README.md:39]: I don't see this command in package.json

File: README.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+Make sure the [Remote Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)
+extension is installed. This method uses Docker Compose where one container is
+the OpenShift console and the second container is the plugin. It requires that
+you have access to an existing OpenShift cluster. After the initial build, the
+cached containers will help you start developing in seconds.
+
+1. Create a `dev.env` file inside the `.devcontainer` folder with the correct values for your cluster:
+
+```bash
+OC_PLUGIN_NAME=kubevirt-plugin
```
Comment (spadgett) [inline README.md:27]: Why kubevirt?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+## Docker image
+
+Before you can deploy your plugin on a cluster, you must build an image and
+push it to an image registry.
+
+1. Build the image:
+
+   ```sh
+   docker build -t quay.io/crontab-plugin/crontab-plugin:latest .
```
Comment (spadgett) [inline README.md:57]: Your quay repository won't be crontab-plugin. It's probably your user ID. We should leave this as a variable in the example.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+oc process -f oc-manifest.yaml \
+  -p PLUGIN_NAME=crontab-plugin \
+  -p NAMESPACE=crontab-plugin \
+  -p IMAGE=quay.io/crontab-plugin/crontab-plugin:latest \
+  | oc create -f -
+```
+
+Once deployed, patch the
+[Console operator](https://github.com/openshift/console-operator)
+config to enable the plugin.
```
Comment (spadgett) [inline README.md:91]: If we use the helm chart, this isn't necessary.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    "typescript": "^4.4.4",
+    "webpack": "^5.68.0",
+    "webpack-cli": "^4.9.1",
+    "webpack-dev-server": "^4.5.0"
+  },
+  "consolePlugin": {
+    "name": "crontab-plugin",
+    "version": "0.0.1",
+    "displayName": "OpenShift CronTab Plugin",
+    "description": "CronTab plugin for OpenShift Console",
```
Comment (spadgett) [inline package.json:66]: Let's improve the description here to make it clear that it's a sample plugin for a custom resource and describe what capabilities it adds.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    "exposedModules": {
+      "CronTabList": "./views/CronTabList/CronTabList.tsx",
+      "CronTabDetails": "./views/CronTabDetails.tsx",
+      "yamlTemplates": "src/templates/index.ts",
+      "modalProvider": "./utils/components/ModalProvider/ModalProvider.tsx",
+      "useCronTabActionsProvider": "./views/hooks/useCronTabActionsProvider.tsx"
+    },
+    "dependencies": {
+      "@console/pluginAPI": "*"
```
Comment (spadgett) [inline package.json:75]: I don't believe it's correct to use `*` anymore, cc @vojtechszocs

File: src/models/CronTabModel.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import { K8sResourceCommon } from '@openshift-console/dynamic-plugin-sdk';
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+
+export const CronTabModel: K8sModel = {
+  label: 'CronTab',
+  labelPlural: 'CronTabs',
+  apiVersion: 'v1',
+  apiGroup: 'stable.example.com',
+  plural: 'crontabs',
+  abbr: 'crontabs',
```
Comment (spadgett) [inline src/models/CronTabModel.ts:11]: `abbr` should be short, like `CT` I think

File: src/models/CronTabModel.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  namespaced: true,
+  kind: 'CronTab',
+  id: 'crontab',
+  crd: true,
+};
+
+export type CronTabKind = K8sResourceCommon & {
```
Comment (spadgett) [inline src/models/CronTabModel.ts:18]: Let's create a better type definition for the actual shape of the resource.

**Review by spadgett — COMMENTED** [review]
Thanks @raspbeep. I haven't looked closely at the code yet, but a few comments on the README and package.json.

File: src/utils/Loading.tsx [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
+
+const Loading: React.FC = () => (
```
Comment (spadgett) [inline src/utils/Loading.tsx:3]: I wouldn't copy in the loader from console, and it's a bad practice for plugins to reference console styles directly since they might change.

We should look at switching console to use the PatternFly loader, and I'd use that here in the plugin (even if different from console today).

File: src/utils/components/ModalProvider/ModalProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,76 @@
+import * as React from 'react';
```
Comment (spadgett) [inline src/utils/components/ModalProvider/ModalProvider.tsx:1]: Was this file copied over from console?

File: src/utils/utils.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+
+const modelToRef = (obj: K8sModel) => `${obj.apiGroup}~${obj.apiVersion}~${obj.kind}`;
```
Comment (spadgett) [inline src/utils/utils.ts:3]: Where do we need to use references? We've been trying to avoid exposing that in the SDK, so we should look at what gap we have.

File: src/utils/hooks/useCronTabTranslation.ts [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * A Hook for using the i18n translation.
+ */
+export const useCronTabTranslation = () => useTranslation('plugin__crontab-plugin');
+
+/**
+ * a function to perform translation to 'plugin__crontab-plugin' namespace
+ * @param value string to translate
+ */
+// skipcq: JS-C1002
```
Comment (spadgett) [inline src/utils/hooks/useCronTabTranslation.ts:12]: What is `skipcq`?

File: src/views/CronTabDetails.tsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        name: t('Details'),
+        component: CronTabDetailsPage,
+      },
+      {
+        href: 'yaml',
+        name: t('YAML'),
+        component: CronTabYAMLPage,
+      },
+    ],
+    [],
```
Comment (spadgett) [inline src/views/CronTabDetails.tsx:48]: I think we need `t` as a dependency here to force re-render when the language changes?

Since others are using this as a model, we should check what's right here.

File: src/views/components/CronTabActions/CronTabActions.scss [hunk 16]
```diff
@@ -0,0 +1,11 @@
+.console-cron-tab-actions {
```
Comment (spadgett) [inline src/views/components/CronTabActions/CronTabActions.scss:1]: We should avoid overriding PF styles. Why is this needed?

File: src/views/CronTabList/CronTabList.scss [hunk 17]
```diff
@@ -0,0 +1,16 @@
+.co-operator-details__actions {
```
Comment (spadgett) [inline src/views/CronTabList/CronTabList.scss:1]: We should avoid overriding console or PF styles. Why is this needed?

File: src/utils/constants.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+export const DEFAULT_NAMESPACE = 'default';
```
Comment (spadgett) [inline src/utils/constants.ts:1]: How is this used?

We should avoid putting anything in the `default` namespace

File: src/views/modals/LabelsModal/LabelsModal.tsx [hunk 19]
```diff
@@ -0,0 +1,158 @@
+import * as React from 'react';
```
Comment (spadgett) [inline src/views/modals/LabelsModal/LabelsModal.tsx:1]: Was this copied over from console? We should avoid reimplementing console functionality and instead open an issue to expose the labels modal.

File: src/views/modals/DeleteModal/DeleteModal.tsx [hunk 20]
```diff
@@ -0,0 +1,83 @@
+import * as React from 'react';
```
Comment (spadgett) [inline src/views/modals/DeleteModal/DeleteModal.tsx:1]: Was this copied over from console? We should avoid reimplementing console functionality and instead open an issue to expose the delete modal.

File: src/views/modals/AnnotationsModal/AnnotationsModal.tsx [hunk 21]
```diff
@@ -0,0 +1,110 @@
+import * as React from 'react';
```
Comment (spadgett) [inline src/views/modals/AnnotationsModal/AnnotationsModal.tsx:1]: Was this copied over from console? We should avoid reimplementing console functionality and instead open an issue to expose the annotations modal.

**Review by spadgett — CHANGES_REQUESTED** [review]
I'm still working through the review, but wanted to get you feedback. We should try to keep the plugin simple and avoid copying content over from openshift/console. If there are missing capabilities, we need to open issues to track. Ideally we should be using PF components and console SDK components with minimal code in the plugin itself.

File: src/utils/utils.ts (hunk 13 above)
Comment (raspbeep) [reply src/utils/utils.ts:3]: It is used for `ListPageCreate` component from `@dynamic-plugin-sdk` which requires `GroupVersionKind` string (in our case `stable.example.com~v1~CronTab`)

Comment (spadgett) [reply src/utils/utils.ts:3]: @raspbeep Please open a bug for this component. We shouldn't be exposing `GroupVersionKind` strings in the API, so we'll want to deprecate this property and replace it with a prop where these values are specified separately.

File: package.json (hunk 8 above)
Comment (raspbeep) [reply package.json:73]: @vojtechszocs would this be fine?
```
  "dependencies": {
    "@console/pluginAPI": "^4.11"
  }
```

File: console-extensions.json [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+  },
+  {
+    "type": "console.context-provider",
+    "properties": {
+      "provider": { "$codeRef": "modalProvider.ModalProvider" },
+      "useValueHook": { "$codeRef": "modalProvider.useModalValue" }
+    }
+  },
+  {
+    "type": "console.navigation/resource-ns",
```
[the remark this answers is not in the document]
Comment (raspbeep) [reply console-extensions.json:21]: What the `type` value should be then?

**Comment by raspbeep** [comment]
@spadgett I removed the DetailsPage and replaced it with pure PF components. I also edited the readme and the description in package.json. The ModalProvider and various Modals(Labels, Annotations and Delete) stay for now and will be replaced once they are properly exposed in the `dynamic-plugin-sdk`.

File: Dockerfile [hunk 23]
```diff
@@ -0,0 +1,10 @@
+FROM docker.io/library/node:16 AS build
```
Comment (spadgett) [inline Dockerfile:1]: We should rebase on the current plugin template, which uses different images.

File: README.md [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+
+It requires OpenShift 4.11.
+
+The CronTab Dynamic Plugin creates new menu entry, routes, list page, details page and a
+default YAML template for a CRD.
+
+## Deployment on cluster
+
+
+### Option 1: Installing the Helm Chart (WIP - does not work yet)
```
Comment (spadgett) [inline README.md:13]: Do we know why it doesn't work? I think we want to solve this before merging.

File: README.md [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+`plugin.image`: Specifies the location of the image containing the plugin, to be deployed
+
+Additional parameters can be specified if desired. Consult the chart [values](charts/openshift-console-plugin/values.yaml) file for the full set of supported parameters.
+
+
+### Option 2:
+You can deploy the plugin to a cluster by applying `oc-manifest.yaml`.
+
+```sh
+oc apply -f oc-manifest.yaml
```
Comment (spadgett) [inline README.md:34]: nitpick: I'd probably rename to `manifest.yaml`.

File: README.md [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+
+In another terminal window, run:
+
+1. `oc login`
+2. `yarn run start-console` (requires [Docker](https://www.docker.com) or [podman](https://podman.io))
+
+## Docker image
+
+Before you can deploy your plugin on a cluster, you must build an image and
+push it to an image registry.
```
Comment (spadgett) [inline README.md:63]: This statement more or less contradicts the instructions at the top of the README.

File: src/utils/hooks/useCronTabTranslation.ts
Comment (spadgett) [inline src/utils/hooks/useCronTabTranslation.ts:1]: Again, let's not use utils and put this is a folder like i18n or similar.

File: src/views/CronTabDetails/CronTabPageTitle.tsx [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+              {t('CronTabs')}
+            </Link>
+          </BreadcrumbItem>
+          <BreadcrumbItem>{t('CronTab Details')}</BreadcrumbItem>
+        </Breadcrumb>
+      </div>
+      <div className="co-m-nav-title co-m-nav-title--detail co-m-nav-title--breadcrumbs">
+        <span className="co-m-pane__heading">
+          <h1 className="co-m-pane__name co-resource-item">
+            <span className="co-m-resource-icon co-m-resource-icon--lg">{'DS'}</span>
```
Comment (spadgett) [inline src/views/CronTabDetails/CronTabPageTitle.tsx:32]: This is a gap in our SDK if we don't provide a way to render a resource title like other pages. We need components to render this rather than hard-coding internal console styles.

File: src/views/CronTabDetails/CronTabPageTitle.tsx [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+        <Breadcrumb className="pf-c-breadcrumb co-breadcrumb">
+          <BreadcrumbItem>
+            <Link to={`/k8s/ns/${namespace || DEFAULT_NAMESPACE}/${cronTabModelRef}`}>
+              {t('CronTabs')}
+            </Link>
+          </BreadcrumbItem>
+          <BreadcrumbItem>{t('CronTab Details')}</BreadcrumbItem>
+        </Breadcrumb>
+      </div>
+      <div className="co-m-nav-title co-m-nav-title--detail co-m-nav-title--breadcrumbs">
```
Comment (spadgett) [inline src/views/CronTabDetails/CronTabPageTitle.tsx:29]: We should avoid any `co-` prefixed classes since these are internal and not public APIs. They can change and break plugins from version to version. If we need to wait for the SDK to expose details page components, we should probably wait, or just use PF components for now.

File: src/views/CronTabDetailsPage.tsx [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+  name: string;
+}> & {
+  obj?: any;
+};
+
+const CronTabDetailsPage: React.FC<CronTabDetailsPageProps> = ({ obj }) => {
+  const { t } = useCronTabTranslation();
+  const { createModal } = useModal();
+
+  const onEditLabels = () => {
```
Comment (spadgett) [inline src/views/CronTabDetailsPage.tsx:33]: We should remove this from the PR and wait until it's available in the SDK.

File: src/views/CronTabList/CronTabList.tsx [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+    isList: true,
+    groupVersionKind: {
+      group: 'stable.example.com',
+      kind: 'CronTab',
+      version: 'v1',
+    },
+    namespaced: true,
+    namespace,
+  });
+  // const { t } = useTranslation();
```
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:38]: Need to update this file to use translations

File: src/views/CronTabList/CronTabList.tsx [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+
+const useCronTabColumns = () => {
+  const columns: TableColumn<K8sResourceCommon>[] = React.useMemo(
+    () => [
+      {
+        title: 'Name',
+        id: 'name',
+        transforms: [sortable],
+        sort: 'metadata.name',
+        props: { className: 'pf-m-width-15' },
```
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:106]: Do we need these widths? By default, won't all columns get the same width?

File: src/views/CronTabList/CronTabList.tsx [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+      </TableData>
+    </>
+  );
+};
+
+const useCronTabColumns = () => {
+  const columns: TableColumn<K8sResourceCommon>[] = React.useMemo(
+    () => [
+      {
+        title: 'Name',
```
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:102]: Needs translation

File: src/views/utils/MetadataLabels/MetadataLabels.tsx [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+    (labels && Object.keys(labels).length > 0) ? 
+      ( 
+        <LabelGroup numLabels={10} className="metadata-labels-group">
+          {Object.keys(labels || {})?.map((key) => {
+            return <Label key={key}>{labels[key] ? `${key}=${labels[key]}` : key}</Label>;
+          })}
+        </LabelGroup>
+      ) :
+      (
+        <span className="metadata-labels-no-labels">No labels</span>
```
Comment (spadgett) [inline src/views/utils/MetadataLabels/MetadataLabels.tsx:19]: Can we use a PF style instead of a custom style?

File: src/views/modals/TabModal/TabModal.tsx [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+    headerText,
+    children,
+    isDisabled,
+    submitBtnText,
+    modalVariant,
+    positionTop = true,
+    submitBtnVariant,
+    titleIconVariant,
+  }) => {
+    const [isSubmitting, setIsSubmitting] = React.useState(false);
```
Comment (spadgett) [inline src/views/modals/TabModal/TabModal.tsx:59]: nitpick:
```suggestion
    const [isSubmitting, setSubmitting] = React.useState(false);
```

File: src/views/modals/TabModal/TabModal.tsx [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+
+    const t = (key: string) => key;
+
+    const closeModal = () => {
+      setError(undefined);
+      setIsSubmitting(false);
+
+      const promise = onClose();
+
+      if (promise) promise?.catch(setError);
```
Comment (spadgett) [inline src/views/modals/TabModal/TabModal.tsx:81]:
```suggestion
      promise?.catch(setError);
```

File: src/views/modals/TabModal/TabModal.tsx [hunk 36]
```diff
@@ -0,0 +1,140 @@
...
+                        <a href={error.href} target="_blank" rel="noreferrer">
+                          {error.href}
+                        </a>
+                      </StackItem>
+                    )}
+                  </Stack>
+                </Alert>
+              </StackItem>
+            )}
+            <StackItem>
```
Comment (spadgett) [inline src/views/modals/TabModal/TabModal.tsx:110]: Do we not let you update any of the fields in this modal?

File: src/views/modals/LabelsModal/utils.ts
Comment (spadgett) [inline src/views/modals/LabelsModal/utils.ts:1]: Let's remove any custom logic for editing labels in this PR.

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabRowActions.tsx:1]: Let's wait until we have SDK support before adding all of the actions.

**Review by spadgett — CHANGES_REQUESTED** [review]
Let's break this up into smaller PRs (list page, details page, actions, etc.) and only merge things when ready. If there are missing things in the SDK, it's better to hold the corresponding PR. This is a sample others will base their plugins on, so we shouldn't merge short-term workarounds for SDK gaps. They'll make their way into other plugin code that might not get updated.

**Comment by spadgett** [comment]
These changes were rolled into #2. Closing...
